### PR TITLE
Created a protoc parser with pre-loaded common protos.

### PR DESCRIPTION
**Pull request by mingzhi**
This PR creates a `protoc` parser pre-loaded with [common protos](https://github.com/googleapis/api-common-protos):
1. we use the pre-compiled descriptors of those common protos in Go packages, and therefore they will be shipped in the linter/test binary;
2. the descriptors of those common protos do not contain source_code_info or dependencies; if those information are desired, users should exclude those common protos and provide theirs instead (via `--proto_path`).

### CODE REVIEW

**Comment by robcapo** [comment]
There's a lot going on here. Can we break this up so we can comment on each of the specific things you're doing separately? I have a few general questions / comments:

1. I think the centralized protoc package makes sense to avoid duplicate logic between testutil and cmd, but I don't think it's necessarily high priority.
2. I'm a bit confused about adding the pre-compiled descriptors to common protos. Does this obviate the need for the submodule? What are the tradeoffs of this versus the way we have it working already? It looks like we have introduced quite a few more dependencies, plus some additional complexity to updating the proto references.
3. Are we sure we want to ship the linter with api-common-protos in it? The linter is a useful tool in its own right, and I'm not sure we should package any protos (aside from stdlib as they're part of protoc) with it. If the reason for this is primarily the convenience of removing the need for a command line flag, there are many ways we can achieve the same thing without such a major change (e.g. urfave/cli supports [environment variables](http://github.com/urfave/cli#values-from-the-environment)).

**Comment by mingzhi** [comment]
> There's a lot going on here. Can we break this up so we can comment on each of the specific things you're doing separately? I have a few general questions / comments:

I will try.

> 
> 1. I think the centralized protoc package makes sense to avoid duplicate logic between testutil and cmd, but I don't think it's necessarily high priority.
> 2. I'm a bit confused about adding the pre-compiled descriptors to common protos. Does this obviate the need for the submodule? What are the tradeoffs of this versus the way we have it working already? It looks like we have introduced quite a few more dependencies, plus some additional complexity to updating the proto references.

Yes, it removes the needs of submodule or as a user/developer, how to get a local copy of the common protos.

Pros:
They are Go packages, so `go` tools work normally, including `install`, `test`, and `get`;
Portability -- they are compiled into the binary;
Compatibility between Github and Google3 -- `update.go` will take care of the dependencies.

Cons:
Increasing the binary size -- however, we can use `go` build tag to exclude the auto-generated `api_common_protos.go` to build a lite version, which does not include those common protos.

Tie:
We will use `gen_imports.go` to sync with [api-common-protos](https://github.com/googleapis/api-common-protos) or manually add more proto packages.

> 3. Are we sure we want to ship the linter with api-common-protos in it? The linter is a useful tool in its own right, and I'm not sure we should package any protos (aside from stdlib as they're part of protoc) with it. If the reason for this is primarily the convenience of removing the need for a command line flag, there are many ways we can achieve the same thing without such a major change (e.g. urfave/cli supports [environment variables](https://github.com/urfave/cli#values-from-the-environment)).

If a user provides those common protos via `--proto_path`, those pre-loaded protos will be ignored and the linter will use the user's.

I am envision the common protos as std library in Go or C++ -- it should just work and users/developers should not worry about where it is.

The environment variable does not improve user's experience -- it essentially the same as using `--proto_path` flag.

**Comment by mingzhi** [comment]
I don't have much experience in breaking up a PR, so I am writing notes to help you understand the changes:
1. `parser/internal/protos` contains common protos, but it is only for packages in `parser`;
2. `parser/internal/protos/gen` which checkouts the api-common-proto repo, parses all proto files, and generates import statements in `parser/internal/protos/api_common_protos.go`; Note that the `api_common_protos.go` can be excluded to compile a lite version without those common protos;
3. `parser/protoc` contains a wrapper of `protoc` with the pre-loaded common protos; I envision that we will add more parsers in the `parser`, e.g., `parser/protoparser` which uses jhump's [`protoparser`](https://github.com/jhump/protoreflect/tree/master/desc/protoparse);
4. changes on the linter and `testutil` to use `parser/protoc`.

Most of changes are newly added files/packages.

File: parser/protoc/parser.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		}
+	}
+	var names []string
+	for key := range s {
+		names = append(names, key)
+	}
+	return names, nil
+}
+
+var importPattern = regexp.MustCompile(`\s*import\s+"([a-zA-Z0-9/\.]+)"\s*;`)
```
Comment (robcapo) [inline parser/protoc/parser.go]: Are `-` or `_` allowed in filenames? Also same issue as before with `(?:^|\s+)` probably being a better way to start.

In general, I'm not crazy about using regexes to pull these out. I think we should try not to do any manual parsing if possible.

Comment (robcapo) [reply parser/protoc/parser.go]: Actually, we always have this issue because of multiline comments
```
/**
OLD CODE FOR REFERENCE

package foo;

import "blah.proto";

message Bar {}

NEW CODE BELOW:
*/

package foo.v2;

import "blah/v2.proto";

message BarV2 {}
```

File: parser/internal/protos/find_proto.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	protoV1 "github.com/golang/protobuf/proto"
+	protoV2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// FindByName returned a FileDescriptorProto if the given name is found
+// in the registered proto files.
+func FindByName(filename string) (*descriptorpb.FileDescriptorProto, error) {
+	d := protoV1.FileDescriptor(filename)
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/find_proto.go]: Go protobuf code generator is still using protobuf v1, which registers file descriptor in proto v1. But we are using proto v2.

File: parser/internal/protos/find_proto.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"io/ioutil"
+
+	protoV1 "github.com/golang/protobuf/proto"
+	protoV2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// FindByName returned a FileDescriptorProto if the given name is found
+// in the registered proto files.
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/find_proto.go]: In the `init` function of the imported packages. For example, see [this](https://github.com/googleapis/go-genproto/blob/6af8c5fc6601ab6b41cd32742a65ce2f5bd9db57/googleapis/api/annotations/http.pb.go#L602) and pre-compiled byte string. So if a package is imported in `api-common-protos.go`, the descriptor is available.

File: parser/internal/protos/gen/gen_imports.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const commonAPIProtosURI = "https://github.com/googleapis/api-common-protos"
+
+func main() {
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/gen/gen_imports.go]: In the file comment of `api-common-protos.go`, there is an instruction (`go run gen/gen_imports.go && go fmt`) on how to re-generate itself.

File: parser/internal/protos/gen/gen_imports.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+var packagePattern = regexp.MustCompile(`\s*option\s+go_package\s*=\s*"([a-zA-Z0-9/\.]+);[a-zA-Z0-9]+";`)
+
+func extractGoPkg(filename string) (string, error) {
+	b, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return "", err
+	}
+	findings := packagePattern.FindStringSubmatch(string(b))
+	if findings == nil {
+		return "", fmt.Errorf("No Go package found in file %q", filename)
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/gen/gen_imports.go]: I think we should refer to [the public version of guides](https://developers.google.com/protocol-buffers/docs/reference/go-generated#package). And in general, the common protos should have a `go_package`, so the generated code can be discovered in godoc.

If not, we can fix it then.

File: parser/protoc/parser.go (hunk 1 above)
Comment (mingzhi) [reply parser/protoc/parser.go]: Good points, but it is safe to add unnecessary dependencies in `protoc`. In this case, it is Okay to search and add `blah.proto` if found to the dependencies.

File: parser/internal/protos/find_proto.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+func FindByName(filename string) (*descriptorpb.FileDescriptorProto, error) {
+	d := protoV1.FileDescriptor(filename)
+	if d == nil {
+		return nil, fmt.Errorf("Proto file %q is not found", filename)
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/find_proto.go]: Done.

File: parser/internal/protos/gen/gen_imports.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	fmt.Fprintln(w, ")")
+}
+
+func gitClone(uri, outDir string) error {
+	cmd := exec.Command("git", "clone", uri, outDir)
+	return cmd.Run()
+}
+
+var packagePattern = regexp.MustCompile(`\s*option\s+go_package\s*=\s*"([a-zA-Z0-9/\.]+);[a-zA-Z0-9]+";`)
```
[the remark this answers is not in the document]
Comment (mingzhi) [reply parser/internal/protos/gen/gen_imports.go]: Done.

**Comment by robcapo** [comment]
I still don't think this is the right solution to the dependency problem. We shouldn't have to tie our linter binary to a specific version of `api-common-protos`. And if we do that, we need to at least include clear version information so that users can easily figure out which version of the protos we're using under the hood.

I also think it's going to be confusing to update them. The mechanics of submodules are well known and well documented, so someone new to the project could easily update the dependency if there's an  update to `api-common-protos`.

**Comment by mingzhi** [comment]
Users can use `--proto_path` to override the preloaded protos, just like what they will do without this. It just adds some convenience by default. If we don't like it in the linter binary, just go ahead and toggle the flag `ExcludeCommomProtos` in the parser.

If later we don't like it at all, just delete the `api_common_proros.go`, and supply the proto path just like what you are proposing in the testing.

In short, users or rule developers do not marry to the preloaded protos. The preloaded one only provides convenience.

It aligns our very original goal -- distribute a binary and it just works.

**Comment by lukesneeringer** [comment]
Closing this as #154 makes it unnecessary.
